qemu/tests/pvpanic.py

```diff
@@ -0,0 +1,141 @@
+    """
+    Prepare or restore environment configuration modification.
+    Operation includds enable/disable kdumpservice, configure
+    unknown_nmi_panic for linux, or modify the register value for
+    windows, in order to trigger nmi or do the clean up work.
+
+    :param test: test object
+    :param params: parameters
+    :vm: target vm
+    :session: session created by loggin the vm
+    :enable_pvpanic: to setup or restore the test environment
+    """
+    timeout = int(params.get("timeout", 360))
+    switch_cmds = []
+    if params.get("os_type") == "linux":
+        # switch kdump service and unknown_nmi_panic
+        if enable_pvpanic:
+            switch_cmds.append(params.get("set_kdump_cmd") %
+                               ('off', 'stop'))
+            switch_cmds.append(params.get("set_nmi_panic_cmd") % '1')
+        else:
+            switch_cmds.append(params.get("set_kdump_cmd") %
+                               ('on', 'restart'))
+            switch_cmds.append(params.get("set_nmi_panic_cmd") % '0')
+    else:
+        # modify the register for windows
+        if enable_pvpanic:
+            switch_cmds.append(params.get("set_NMICrashDump_cmd") %
+                               ('SetDwordValue', 'uValue=1'))
+        else:
+            switch_cmds.append(params.get("set_NMICrashDump_cmd") %
+                               ('DeleteValue', ''))
+
+    for cmd in switch_cmds:
+        status, output = session.cmd_status_output(cmd)
+        if status:
+            test.error("Command '%s' failed, status: %s, output: %s" %
+                       (cmd, status, output))
+    if params.get("os_type") == "windows":
+        session = vm.reboot(session, timeout=timeout)
+
+
+def check_qmp_events(vm, event_name, timeout=600):
+    """
+    Check whether certain qmp event appeared in vm.
+
+    :param vm: target virtual machine
+    :param event_name: target event name, such as 'GUEST_PANICKED'
+    :param timeout: check time
+    """
+
+    end_time = time.time() + timeout
+    qmp_monitors = vm.get_monitors_by_type("qmp")
+    qmp_num = len(qmp_monitors)
+    logging.info("Try to get qmp events %s in %s seconds!" %
+                 (event_name, timeout))
+    while time.time() < end_time:
+        for monitor in qmp_monitors:
+            event = monitor.get_event(event_name)
+            if event:
+                logging.info("Receive qmp %s event notification",
+                             event_name)
+                qmp_num -= 1
+                qmp_monitors.remove(monitor)
+        if qmp_num == 0:
+            return True
+        time.sleep(5)
+    return False
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    pvpanic test
+
+    1) Log into the guest
+    2) Check if the driver is installed and verified (only for win)
+    3) Stop kdump service and modify unknown_nmi_panic(for linux)
+       or modify register value(for win)
+    4) Trigger a crash by nmi
```

Comment:
I suggest write the guest crash method into a function.
As we do have several crash methods,nmi is only one of them.

Reply:
updated, pls help to review , thanks